src/charm.py

```diff
@@ -12,92 +12,292 @@
     https://discourse.charmhub.io/t/4208
 """
 
+import itertools
 import logging
+from functools import wraps
+from resource import (
+    ClusterRole,
+    ClusterRoleBinding,
+    ConfigMap,
+    DaemonSet,
+    Deployment,
+    Resource,
+    Role,
+    RoleBinding,
+    Secret,
+    Service,
+    ServiceAccount,
+    StorageClass,
+)
+from typing import Any, Callable, Dict, List
 
-from ops.charm import CharmBase
-from ops.framework import StoredState
+import yaml
+from jinja2 import Environment, FileSystemLoader
+from kubernetes import client, config, utils
+from kubernetes.client.exceptions import ApiException
+from ops.charm import (
+    CharmBase,
+    ConfigChangedEvent,
+    InstallEvent,
+    RelationDepartedEvent,
+    RelationJoinedEvent,
+)
+from ops.framework import StoredDict, StoredState
 from ops.main import main
-from ops.model import ActiveStatus
+from ops.model import ActiveStatus, BlockedStatus
 
 logger = logging.getLogger(__name__)
 
 
+def needs_leader(func: Callable) -> Callable:
+    """Ensure that function with this decorator is executed only if on leader units."""
+
+    @wraps(func)
+    def leader_check(self: CharmBase, *args: Any, **kwargs: Any) -> Any:
+        if self.unit.is_leader():
+            func(self, *args, **kwargs)
```

Comment:
Could this function log when `Unit is not the leader. Skipping execution of {func}`

Reply:
I was thinking about this, but there's not really a good way to get "human readable" name of the function. We could use `func.__name__` but that really exposes internal implementation and might not be easy to understand without knowing context (code).
We can certainly do it, but I thought that it does add that much value to have non-leader units announce that the are NOT doing something :D

Reply:
leadership may change for different reasons. A unit may show as leader but have logs about not being the leader (from a few minutes or hours ago).

OTOH, support and operations teams use juju-crashdump which collect all the Juju logs from a Juju model. In debug mode, Juju will tell if a unit is the leader or not. But I think it would not hurt to make the unit itself log about it, which from my experience is more useful than the Juju modules logs.